solver/llbsolver/history.go

```diff
@@ -1028,6 +1037,199 @@ func (h *HistoryQueue) Listen(ctx context.Context, req *controlapi.BuildHistoryR
+	for _, in := range in {
+		fns, err := parseFilter(in)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, func(ev *controlapi.BuildHistoryEvent) bool {
+			for _, fn := range fns {
+				if !fn(ev) {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	return out, nil
+}
+
+func timeBasedFilter(f string) (func(*controlapi.BuildHistoryEvent) bool, error) {
+	key, sep, value, _ := cutAny(f, []string{">=", "<=", ">", "<"})
+	var cmp int64
+	switch key {
+	case "startedAt", "completedAt":
+		v, err := time.ParseDuration(value)
+		if err == nil {
+			tm := time.Now().Add(-v)
+			cmp = tm.Unix()
+		} else {
+			tm, err := time.Parse(time.RFC3339, value)
+			if err != nil {
+				return nil, errors.Errorf("invalid time %s", value)
+			}
+			cmp = tm.Unix()
+		}
+	case "duration":
+		v, err := time.ParseDuration(value)
+		if err != nil {
+			return nil, errors.Errorf("invalid duration %s", value)
+		}
+		cmp = int64(v)
+	default:
+		return nil, nil
+	}
+
+	return func(ev *controlapi.BuildHistoryEvent) bool {
+		if ev.Record == nil {
+			return false
+		}
+		var val int64
+		switch key {
+		case "startedAt":
+			val = ev.Record.CreatedAt.AsTime().Unix()
+		case "completedAt":
+			if ev.Record.CompletedAt != nil {
+				val = ev.Record.CompletedAt.AsTime().Unix()
+			}
+		case "duration":
+			if ev.Record.CompletedAt != nil {
+				val = int64(ev.Record.CompletedAt.AsTime().Sub(ev.Record.CreatedAt.AsTime()))
+			}
+		}
+		switch sep {
+		case ">=":
+			return val >= cmp
+		case "<=":
+			return val <= cmp
+		case ">":
+			return val > cmp
+		default:
+			return val < cmp
+		}
+	}, nil
+}
+
+func parseFilter(in string) ([]func(*controlapi.BuildHistoryEvent) bool, error) {
+	var out []func(*controlapi.BuildHistoryEvent) bool
+
+	fields, err := csvvalue.Fields(in, nil)
+	if err != nil {
+		return nil, err
+	}
+	var staticFilters []string
+
+	for _, f := range fields {
+		fn, err := timeBasedFilter(f)
+		if err != nil {
+			return nil, err
+		}
+		if fn == nil {
+			staticFilters = append(staticFilters, f)
+			continue
+		}
+		out = append(out, fn)
+	}
+
+	filter, err := filters.ParseAll(strings.Join(staticFilters, ","))
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse history filters %v", in)
+	}
+
+	out = append(out, func(ev *controlapi.BuildHistoryEvent) bool {
+		if ev.Record == nil {
+			return false
+		}
+		return filter.Match(adaptHistoryRecord(ev.Record))
+	})
+	return out, nil
+}
+
+func adaptHistoryRecord(rec *controlapi.BuildHistoryRecord) filters.Adaptor {
+	return filters.AdapterFunc(func(fieldpath []string) (string, bool) {
+		if len(fieldpath) == 0 {
+			return "", false
+		}
+
+		switch fieldpath[0] {
+		case "ref":
+			return rec.Ref, rec.Ref != ""
+		case "status":
+			if rec.CompletedAt != nil {
+				if rec.Error != nil {
+					if strings.Contains(rec.Error.Message, "context canceled") {
+						return "canceled", true
+					}
+					return "error", true
+				}
+				return "complete", true
```

Comment:
I think it should be `completed` so we are aligned with https://github.com/docker/buildx/pull/2964/files#diff-19d235e8c696e21245207f6a179e939adb7a3fb12fd8830d75dc134e5a1bf0dfR52

Can we have the status as public consts?